Script hierarchies for Inspector

willnationsdev opened the pull request:
Fixes #10799 

Demonstration:

Here are some various screenshots. These are from an environment in which I have the following structure:

    # class_c.gd
    extends "res://class_b.gd"
    export(bool) var c = true
    export(bool) var c2 = false
    
    # class_b.gd
    extends "res://class_a.gd"
    var b = true
    
    # class_a.gd
    extends "res://class_2.gd"
    export var a = "a"
    export var a2 = "a2"
    
    # class_2.gd
    extends "res://class_1.gd"
    export var three = 3
    export var four = 4
    
    # class_1.gd
    extends Node
    export var one = 1
    export var two = 2

The only variable/class we should NOT see show up is `class_b.gd` with the single, un-exported `b` property with value `true`.

If I attach `class_c.gd` to a Node, I see this before/after:

![class_c_before](https://user-images.githubusercontent.com/16217563/34071983-f3f6c350-e244-11e7-803c-85046580828a.PNG)
![class_c_after](https://user-images.githubusercontent.com/16217563/34071939-464641cc-e244-11e7-808c-4bc14bff6155.PNG)

If I have a builtin script with exported `builtin1` and non-exported `builtin2`, then I get this before/after (`script_test.tscn` is the name of the scene and the built-in script extends a `Node` type):

![builtin_script_before](https://user-images.githubusercontent.com/16217563/34071953-5a8c3182-e244-11e7-9194-b969c4976b17.PNG)
![builtin_script_after](https://user-images.githubusercontent.com/16217563/34071955-60a4e1fe-e244-11e7-8c16-cfe749116b81.PNG)

If I have `class_c` open, and I change it to extend from `class_2` instead of `class_b`, but I HAVEN'T saved it yet, then I will see this before/after (looks the same as the saved version, which was the goal):

![derive_2_not_b_unsaved_c_before](https://user-images.githubusercontent.com/16217563/34071961-7f72b336-e244-11e7-9e45-b2f5a74a13a7.PNG)
![derive_2_not_b_unsaved_c_after](https://user-images.githubusercontent.com/16217563/34071963-83ce9044-e244-11e7-93df-885e61020f3d.PNG)

~~In the course of doing this, I did find another small bug for the unsaved scripts (going from less derived to more derived base class in an unsaved script. In this case, I went from `class_1` to `class_a`, so `class_2` and `class_a's variables are showing up under `class_1`'s), so it seems it still needs some more work.~~

Edit: Oh wait! No, that's a purposeful "bug". If you ADD scripts and properties to the inheritance hierarchy, but the script hasn't been saved, the `p_list` ends up with all of these extra properties and doesn't yet know about the new scripts, so it just bundles them all underneath the most relevant script it can find. It gets revised when you actually save the script.

![derive_a_not_1_unsaved_c_after](https://user-images.githubusercontent.com/16217563/34072091-d5a3f3d0-e246-11e7-96e0-07fefc8d23c0.PNG)
(after saving:)
![class_c_after](https://user-images.githubusercontent.com/16217563/34072148-8ede8266-e247-11e7-9a86-cd2fa65a506b.PNG)

Incidentally, if you add a script to a node that has nothing else in the script hierarchy, and the script is not yet saved (so it doesn't have a name yet), AND if it has export variables that show up in the Inspector, then it starts autonaming the category blocks with "Class 0", "Class 1", etc. for every script that isn't yet saved. If anywhere in the script hierarchy there is a saved script, then all of the unsaved scripts' properties are assumed to belong to the saved script, until those other scripts are saved. This is why we see the "buggy" behavior above. Here's what this algorithm results in for the objects with no saved scripts:

![unsaved_new_script_export](https://user-images.githubusercontent.com/16217563/34072268-c8464fd2-e249-11e7-8569-1ecf3283d0ad.PNG)

Regardless of whether this ends up as part of the 3.0 release or not, I might actually start extending this out into a full on overhaul of custom types in general, targeting a 3.1 release. This would be a critical change to have for a full custom type simulation across the engine.

Edit:

Here are some various screenshots. These are from an environment in which I have the following structure:

    # class_c.gd
    extends "res://class_b.gd"
    export(bool) var c = true
    export(bool) var c2 = false
    
    # class_b.gd
    extends "res://class_a.gd"
    var b = true
    
    # class_a.gd
    extends "res://class_2.gd"
    export var a = "a"
    export var a2 = "a2"
    
    # class_2.gd
    extends "res://class_1.gd"
    export var three = 3
    export var four = 4
    
    # class_1.gd
    extends Node
    export var one = 1
    export var two = 2

The only variable/class we should NOT see show up is `class_b.gd` with the single, un-exported `b` property with value `true`.

If I attach `class_c.gd` to a Node, I see this before/after:

![class_c_before](https://user-images.githubusercontent.com/16217563/34071983-f3f6c350-e244-11e7-803c-85046580828a.PNG)
![class_c_after](https://user-images.githubusercontent.com/16217563/34071939-464641cc-e244-11e7-808c-4bc14bff6155.PNG)

If I have a builtin script with exported `builtin1` and non-exported `builtin2`, then I get this before/after (`script_test.tscn` is the name of the scene and the built-in script extends a `Node` type):

![builtin_script_before](https://user-images.githubusercontent.com/16217563/34071953-5a8c3182-e244-11e7-9194-b969c4976b17.PNG)
![builtin_script_after](https://user-images.githubusercontent.com/16217563/34071955-60a4e1fe-e244-11e7-8c16-cfe749116b81.PNG)

If I have `class_c` open, and I change it to extend from `class_2` instead of `class_b`, but I HAVEN'T saved it yet, then I will see this before/after (looks the same as the saved version, which was the goal):

![derive_2_not_b_unsaved_c_before](https://user-images.githubusercontent.com/16217563/34071961-7f72b336-e244-11e7-9e45-b2f5a74a13a7.PNG)
![derive_2_not_b_unsaved_c_after](https://user-images.githubusercontent.com/16217563/34071963-83ce9044-e244-11e7-93df-885e61020f3d.PNG)

~~In the course of doing this, I did find another small bug for the unsaved scripts (going from less derived to more derived base class in an unsaved script. In this case, I went from `class_1` to `class_a`, so `class_2` and `class_a's variables are showing up under `class_1`'s), so it seems it still needs some more work.~~

Edit: Oh wait! No, that's a purposeful "bug". If you ADD scripts and properties to the inheritance hierarchy, but the script hasn't been saved, the `p_list` ends up with all of these extra properties and doesn't yet know about the new scripts, so it just bundles them all underneath the most relevant script it can find. It gets revised when you actually save the script.

![derive_a_not_1_unsaved_c_after](https://user-images.githubusercontent.com/16217563/34072091-d5a3f3d0-e246-11e7-96e0-07fefc8d23c0.PNG)
(after saving:)
![class_c_after](https://user-images.githubusercontent.com/16217563/34072148-8ede8266-e247-11e7-9a86-cd2fa65a506b.PNG)

Incidentally, if you add a script to a node that has nothing else in the script hierarchy, and the script is not yet saved (so it doesn't have a name yet), AND if it has export variables that show up in the Inspector, then it starts autonaming the category blocks with "Class 0", "Class 1", etc. for every script that isn't yet saved. If anywhere in the script hierarchy there is a saved script, then all of the unsaved scripts' properties are assumed to belong to the saved script, until those other scripts are saved. This is why we see the "buggy" behavior above. Here's what this algorithm results in for the objects with no saved scripts:

![unsaved_new_script_export](https://user-images.githubusercontent.com/16217563/34072268-c8464fd2-e249-11e7-8569-1ecf3283d0ad.PNG)

Regardless of whether this ends up as part of the 3.0 release or not, I might actually start extending this out into a full on overhaul of custom types in general, targeting a 3.1 release. This would be a critical change to have for a full custom type simulation across the engine.

willnationsdev commented:
List::insert_before and List::insert_after used to not properly hook up the previous / following node to the new neighbor. Had been running into problems doing a forward loop after calling `insert_before`, but now it works.

akien-mga commented on core/object.cpp:
```diff
@@ -435,93 +435,212 @@ void Object::set(const StringName &p_name, const Variant &p_value, bool *r_valid
-		if (r_valid)
-			*r_valid = true;
-		return;
+metadata = p_value;
```
You broke indentation all over the file :)
See http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#using-clang-format-locally

willnationsdev replied:
oh gosh. Yeah, thanks for this. Hopefully I can fix it up. XD

willnationsdev commented:
I'm gonna have to make some modifications to this because I've encountered a bug where my code is crashing the engine.

akien-mga commented:
It might be worth splitting your `List` fix in a separate commit and another PR so that it can be reviewed seprately.

willnationsdev commented:
@akien-mga Is that a good idea if the code relies on it? I mean, if I separate it out, I won't be able to test the script changes locally without changing the code to not use `insert_before` at all. Would have to re-organize it into `push_back` uses and re-implement everything all over again.

willnationsdev commented:
I suppose I could separate it and then keep the change, but just not have it recorded by git in the script-var-categories branch...

akien-mga commented:
You can make that commit in a new branch and a new PR, and rebase this PR on that other branch (so you would have two commits).

willnationsdev commented:
oh, so that sounds good. Will get on that.

akien-mga commented:
@willnationsdev Could you now rebase again since #13143 was merged but the related commit here has a different hash? (also you might want to remove the `+` of the commit log ;))

willnationsdev commented:
Yes, I can do that later. I will likely be away from a computer for the next day or two. Will ping you when I have it ready.

akien-mga commented:
No worry, I can also merge it manually myself when it's approved if need be.

willnationsdev commented:
That works too. I can ping you early if I get a chance to work on it sooner rather than later. Otherwise, I'll see the merge notification. Thanks! :-)

willnationsdev commented:
I still won't get to work on this until later, but keep in mind that currently, there seems to be an issue when you have a single script with only one exported variable. Something is causing the `script_vars_start` variable to be null in some cases.

willnationsdev committed:
Script Hierarchies in Inspector

willnationsdev commented:
This latest version should work "perfectly" since I tested it with a lot of bizarre cases and it now properly accounts for unsaved scripts that have updated export properties (changed "extends", added/removed/renamed exports, etc.) by putting these items under separate anonymous category dividers until the script is saved.

willnationsdev commented:
Ok, I've run clang-format, rebased onto master, and removed the now unnecessary List fix commit. If the checks complete safely, then this PR should be good to go.

willnationsdev commented:
I've realized that the code I've written only displays the default name in cases where the script is a GDScript that is built-in to the scene file. I can try to make an adjustment for that, potentially.

willnationsdev commented:
Ok, a solution for this has also been made now. They now have names with the structure...

[capitalized scene file name]:[in-engine class name]:[built-in script index for this property list].

So a scene at "res://family.tscn" with...

- parent (Node, built-in GDScript)
--me (Node, built-in GDScript)
---child (Node, built-in VisualScript)

would result in the category names

- Family:Node:0, Node, Script, etc.
-- Family:Node:0, Node, Script, etc.
--- Family:Node:0, node, Script, etc.

akien-mga commented:
To be honest I'm not sure to fully understand what this does. I thought the intent was just to categorize export vars based on what script declares them in the inheritance path, but it's a lot of code in the end, I'm a bit puzzled.

I think it's probably too late to merge such a change for 3.0, but it will be properly reviewed for 3.1. In the meantime, it would be nice to update the OP with more details about what this is doing, screenshots of the before/after, etc.

willnationsdev commented:
The reason there is a lot of code is for several reasons.

1. the Script class doesn't provide a simple means of informing you about the number of exported variables, so you instead have to parse through the `get_script_property_list`, and filter out all of the variables that aren't exported.

2. Because the property itself doesn't know to which script it belongs and because inherited scripts can't distinguish between their own variables and their parents' variables, you have to grab property lists for every single script in the hierarchy and march through the live, Inspector-used `p_list` and compare the properties' names to see which properties belong to which script.

3. The 2nd process is made more complex by the fact that if a new script hasn't been saved yet, if exported variables have been added, removed, or renamed, or if the inheritance hierarchy for an unsaved script has been modified, then the script and/or its variables won't even exist in `p_list`, so you have to come up with a solution of where to position those mismatching properties that don't exist in the property lists from the list of scripts.

4. You also have to devise a solution for saved and unsaved built-in scripts which, according to the Object, don't have a script from which to get a name. As such, I have to devise the name from information like the scene file name, and the engine type of the Object.

I have been doing extensive testing for this on my own Windows machine (hence the 4 or 5 different iterations I had to go through to get it ready).

willnationsdev commented:
In a few minutes, I'll post some various screenshots. These are from an environment in which I have the following structure:

    # class_c.gd
    extends "res://class_b.gd"
    export(bool) var c = true
    export(bool) var c2 = false
    
    # class_b.gd
    extends "res://class_a.gd"
    var b = true
    
    # class_a.gd
    extends "res://class_2.gd"
    export var a = "a"
    export var a2 = "a2"
    
    # class_2.gd
    extends "res://class_1.gd"
    export var three = "3"
    export var four = "4"
    
    # class_1.gd
    extends Node
    export var one = 1
    export var two = 2

The only variable/class we should NOT see show up is `class_b.gd` with the single, un-exported `b` property with value `true`.

If I attach `class_c.gd` to a Node, I see this before/after:

![class_c_attached_before](https://user-images.githubusercontent.com/16217563/34071658-4594364e-e23f-11e7-9a62-1c477ceb3889.PNG)
![class_c_after](https://user-images.githubusercontent.com/16217563/34071939-464641cc-e244-11e7-808c-4bc14bff6155.PNG)

If I have a builtin script with exported `builtin1` and non-exported `builtin2`, then I get this before/after:

![builtin_script_before](https://user-images.githubusercontent.com/16217563/34071953-5a8c3182-e244-11e7-9194-b969c4976b17.PNG)
![builtin_script_after](https://user-images.githubusercontent.com/16217563/34071955-60a4e1fe-e244-11e7-8c16-cfe749116b81.PNG)

If I have `class_empty` open, and I change it to extend from `class_2` instead of `class_b`, but I HAVEN'T saved it yet, then I will see this before/after (looks the same as the saved version, which was the goal):

![derive_2_not_b_unsaved_c_before](https://user-images.githubusercontent.com/16217563/34071961-7f72b336-e244-11e7-9e45-b2f5a74a13a7.PNG)
![derive_2_not_b_unsaved_c_after](https://user-images.githubusercontent.com/16217563/34071963-83ce9044-e244-11e7-93df-885e61020f3d.PNG)

In the course of doing this, I did find another small bug for the unsaved scripts (going from less derived to more derived base class in an unsaved script, `class_a`'s variables are showing up under `class_1`'s), so it seems it still needs some more work.

If this is going to be part of the 3.1 release, then I might actually start extending this out into a full on overhaul of custom types in general. This change would be a critical change to have for a full custom type simulation across the engine.

willnationsdev commented:
I'm removing the built-in script ":[index]" annotation for the category block now that I've discovered that you can't derive from a built-in script by pointing to the scene file. Now the built in script will only show `[scene_file_name]:[engine_type_name]`.

![updated_builtin](https://user-images.githubusercontent.com/16217563/34072237-200e5eea-e249-11e7-80c7-8d0b72a60b04.PNG)

toger5 commented:
How hard would it be to Introduce proper coustom class names.
So you can do 'extends MyCoustomClass' instead of: 'extends "res://path/to/file/myCoustomClass.gd"'

willnationsdev commented:
@toger that's actually a project I've been working on for a while. Been trying to unify the custom type system API with the ability to register namespaces and typenames for scripts. There is a TypeDB issue I created about it before.

Basically, once it's done, you should be able to do `extends Scripts.namespace.typename`.

toger5 commented:
Nice!!! This is amazing and something i was hoping for since a long time.
How far could you go with providing type information? Would inheriting from a scene let the intellisense know about the node structure in it?

willnationsdev commented:
I don't really know how the Intellisense works (haven't looked at it), but it's theoretically possible. All I will really be doing under the hood is keeping a HashMap of namespaces typenames to file paths for each resource. It will be a slash-delimeted typename mapped to a file path, stored in ProjectSettings. Then I am planning to detect the use of "Scripts" or "Scenes", begin parsing a dot-delimited set of names, take the whole of those names, replacing the dots with slashes, and then plugging it in to get the path to the appropriate resource, returning the loaded resource. This keeps us from having to constantly keep custom type resources loaded in memory.

Now, if the resource were a scene, then you would be able to get the SceneState of the PackedScene, analyze it's contents, and then fill an auto completion popup with the node paths in the scene to somehow get a reference to that node. You'd probably have to store the SceneState alongside the variable data so that it'd be accessible when attempting.pt to get a node from that variable.

toger5 commented:
Makes sense. I thought the custom type would become part of the classDB.
I think this is what the intellisense right now uses for querying functions/properties.

Are there anybissues when saving the coustom type to file connection in the ProjectSettings when moving files from one project to another to reuse them?

How hard would it be to use classDB.
Class db also could add the advantage of letting godot create a docfile for all your classes api. (Even functikn descriptikns could be added via comment. Which would be amazing for plugin documentation.)

willnationsdev commented:
@toger5 

> I thought the custom type would become part of the classDB. I think this is what the intellisense right now uses for querying functions/properties.

> How hard would it be to use classDB.

I would actually be much easier to use the ClassDB, and in fact, that's what I did at first. But reduz didn't want what should really only be editor-time type checking (for custom type reinforcement) and GDScript-module-specific parser/compiler changes affect the core of the engine (rightly so). So I refactored all the code out of the ClassDB and inserted it into a TypeDB in the EditorData class. This keeps it globally accessible through the EditorNode singleton, but keeps all of the dependencies strictly in the Editor or language-specific module space.

I've actually refactored the whole thing about 3 different times. That's the only reason it's taken so long to get this far. Started this project around the beginning of the year.

> Are there anybissues when saving the coustom type to file connection in the ProjectSettings when moving files from one project to another to reuse them?

There are no issues with moving custom types. The EditorNode has a callback for when the filesystem is changed. It catalogues which files are new / have been modified, and I pass those files to the TypeDB which re-registers any existing files with updated typenames / file locations if those ever change.

> Class db also could add the advantage of letting godot create a docfile for all your classes api. (Even functikn descriptikns could be added via comment. Which would be amazing for plugin documentation.)

Yes, this would be great! And luckily, as it turns out, you don't need the ClassDB to do any of that. There is a DocData class associated with the Editor that allows you to search a directory for XML files and generate documentation from those files. I plan on exposing this method to the EditorPlugin class so that users can directly have their plugin generate updated documentation for their content.

toger5 commented:
this all sounds really promising! Thank you! for taking the time to give me an insight on how it works.

>  I plan on exposing this method to the EditorPlugin class so that users can directly have their plugin generate updated documentation for their content.

maybe a second function for parsing doc information to DocData could be created which parses it directly from the script file itself. with custom comments which than represent the documentation for a fucniton/member. how feasible do you think this would be?

This Is also discussed in a different issue?

willnationsdev commented:
@toger5 

> maybe a second function for parsing doc information to DocData could be created which parses it directly from the script file itself. with custom comments which than represent the documentation for a fucniton/member. how feasible do you think this would be?

You are probably looking for the annotations feature that has been discussed somewhere (I can't seem to find the related Issue(s)?). At one point, we were throwing around the idea of adding @-based annotations in GDScript so that you could add metadata about class content for documentation and autocompletion features. So, something like...

@param p_a: variant, The first parameter.
@param p_b: variant, The second parameter.
@return variant, The sum.
func add(p_a, p_b):
----return p_a + p_b

willnationsdev commented:
@mhilbrunner Don't C# scripts allow for inheritance too? I may have to test this with a C# script as well, but I haven't messed around with that at all.

neikeq commented:
get_base_script() is not yet implemented in C#.

willnationsdev commented:
@neikeq ah, thanks. I'll have to do that.

mhilbrunner commented:
As discussed in the IRC meeting, the idea behind this is good and welcome, but needs to be implemented in a solid way.

Notes from IRC:
```
reduz> I still think bojidar_bgs PR for script categories was the best solution to this problem
too bad it was closed
the approach was bad, but the idea was good
a keyword export_category "Something" would be great
Akien> I think we're starting to have a need for annotations support in GDScript finally :)
Akien>We could do stuff like:
#[category "Something"]
export var foo
```
